### PR TITLE
Add option to scan for an existing container

### PR DESCRIPTION
**Pull request by legionus**
As follow up of #55 I added the ability to scan existing containers. All options are applicable to the image (such as `-scan-type`, `-post-results-url` etc.) can be applied to the scanning of the container. 

```
$ ./image-inspector -container 37017548dc46 -scan-type clamav -clam-socket /var/run/clamd.scan/clamd.sock 
2017/07/19 14:47:49 Extracting image  to /var/tmp/image-inspector-434280052
2017/07/19 14:47:51 clamav scan took 0s (1 problems found)
```

### CODE REVIEW

**Comment by simon3z** [comment]
@ilackarms can you review?

File: pkg/inspector/image-inspector.go [hunk 1]
```diff
@@ -359,7 +372,7 @@ func (i *defaultImageInspector) pullImage(client *docker.Client) error {
 // option's destination path.  If the destination path is empty it will write to a temp directory
 // and update the option's destination path with a /var/tmp directory.  /var/tmp is used to
 // try and ensure it is a non-in-memory tmpfs.
-func (i *defaultImageInspector) createAndExtractImage(client *docker.Client, containerName string) (*docker.Image, error) {
+func (i *defaultImageInspector) createAndExtractImage(client *docker.Client, containerName string) (*docker.Image, *docker.Container, error) {
```
Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: the only place where `createAndExtractImage` is being called is https://github.com/openshift/image-inspector/pull/66/files#diff-cb14c05f76ccda787aced81d2a624e69R153 where we throw away the `*docker.Container` value. Why return it at all?

**Comment by ilackarms** [comment]
overall change looks good to me other than one nit.

however I would like to see tests added for this change

File: pkg/cmd/types.go [hunk 2]
```diff
@@ -90,8 +92,11 @@ func (i *ImageInspectorOptions) Validate() error {
 	if len(i.URI) == 0 {
 		return fmt.Errorf("docker socket connection must be specified")
 	}
-	if len(i.Image) == 0 {
-		return fmt.Errorf("docker image to inspect must be specified")
+	if len(i.Image) > 0 && len(i.Container) > 0 {
+		return fmt.Errorf("options -container and -image are mutually exclusive")
```
Comment (pweil-) [inline pkg/cmd/types.go]: remove - from the names

File: cmd/image-inspector.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ func main() {
 
 	flag.StringVar(&inspectorOptions.URI, "docker", inspectorOptions.URI, "Daemon socket to connect to")
 	flag.StringVar(&inspectorOptions.Image, "image", inspectorOptions.Image, "Docker image to inspect")
+	flag.StringVar(&inspectorOptions.Container, "container", inspectorOptions.Container, "Docker container to inspect")
```
Comment (pweil-) [inline cmd/image-inspector.go]: add help text saying one or the other must be specified

File: pkg/inspector/image-inspector.go (hunk 1 above)
Comment (legionus) [inline pkg/inspector/image-inspector.go]: @ilackarms Fixed.

File: pkg/cmd/types.go (hunk 2 above)
Comment (legionus) [inline pkg/cmd/types.go]: @pweil- Done

File: cmd/image-inspector.go (hunk 3 above)
Comment (legionus) [inline cmd/image-inspector.go]: @pweil- Added.

**Comment by legionus** [comment]
> Is there anything we have to worry about if a live container is changing while it's being extracted?

The scan will occur periodically. So I'm OK if the container will change. This is a payment for the fact that we do not stop the container.

> One of the goals is to be able to scan the mounts of the container too. Did you have thoughts on how that would happen or could be integrated into the process?

I checked. All volumes mounted in the container will be scanned. The `DownloadFromContainer` does not remain within a one filesystem.

**Comment by pweil-** [comment]
> I checked. All volumes mounted in the container will be scanned. The DownloadFromContainer does not remain within a one filesystem.

When you say that all volumes will be scanned does that mean that the extract from a running container is extracting the mounted volumes (I wouldn't have thought that was the case) or that they're being scanned outside the scope of this change?

**Comment by legionus** [comment]
> When you say that all volumes will be scanned does that mean that the extract from a running container is extracting the mounted volumes (I wouldn't have thought that was the case)

Yes. All container volumes will be extracted with the root filesystem.

**Comment by pweil-** [comment]
> Yes. All container volumes will be extracted with the root filesystem.

@simon3z We should have a discussion about this with the group and see what safeguards we may need in place if it's going to extract mounted volumes.

**Comment by joelsmith** [comment]
I'm still not convinced that copying is the right approach. I like that the code is simple and straightforward, but it seems terribly inefficient to me.

* A container that has a PV with a multi-gigabyte database would require a complete copy for every scan
* A container with a large 1 G log file written to the ephemeral layer would require a complete copy for every scan

In the case of a mounted volume, my preference would be for us to determine the local path to the mounted volume, and scan that directly, then map the host path to the volume path in the results.

In the case of the container (ephemeral) layer, my preference is to do a docker diff, then re-mount the root volume (only) and scan the changed files. Especially in light of the fact that we'll soon be able to switch to OverlayFS, this scan method will be very efficient.

**Comment by legionus** [comment]
This means that this PR is completely wrong.

**Comment by joelsmith** [comment]
Sorry, I didn't mean to end the discussion.

I know there are issues with remounting root volumes of containers that use the DM storage driver. So I think it's possible that after discussion, we might decide that it's worth paying the price of the copies to avoid those issues. Maybe we'd plan to replace the copy code with a re-mount approach once we get to OverlayFS.

**Comment by pweil-** [comment]
Yes, let's continue the discussion and design, this PR is moving us in the right direction but we need a little thought on what it means to copy everything.

**Comment by legionus** [comment]
@pweil- I thought to offer another PR in which to try to avoid `DownloadFromContainer` at all. We discussed with @joelsmith it before, and I also don't like the idea of copy everything.

**Comment by legionus** [comment]
Please take a look on PoC. Now container scanning does not require copying all the content.

These changes require https://github.com/openshift/clam-scanner/pull/3

**Comment by joelsmith** [comment]
This looks great! I think it's great that you're noticing when the container goes away so that you can interrupt the scan and minimize the issues of keeping a DM device mounted that docker would be trying to unmount.

For a container scan, I think it would be good if the scanner had the ability only to scan files that have changed relative to the image. The assumption would be that the image has already been scanned, so we can save a lot of time and resources by only scanning changed or added files.
For files returned by the call
https://docs.docker.com/engine/api/v1.30/#operation/ContainerChanges
only files with a 0 or 1 (modified or added) need to be scanned.

While I'm most interested in the use case of scanning added and changed files, it might still be useful to keep the ability to scan the whole container, so if you want to make an option for scanning only changes, but make the current behavior the default, I think that would be fine.

I'm surprised that this code works without ever specifying the "remount" option. As I recall, I had to specify the "remount" option or Linux would return an error when I tried to mount something that was already mounted by a container.

Also, I think it's a good idea to add the "ro" option so that we don't risk anything ever modifying the container's fs while a scan is happening.

Overall, it's looking really good.

**Comment by legionus** [comment]
> For a container scan, I think it would be good if the scanner had the ability only to scan files that have changed relative to the image.

In this case I should not mount rootfs, but back to copying ((

I thought I'd make a filter to list the files that need to scan, but unfortunately OSCAP takes a destDir as an arguments. There is no such problem with CLAMAV.

**Comment by joelsmith** [comment]
FYI to anyone following this PR, @legionus and I had a conversation about this on IRC. His idea is that scanners can just ignore the option of scanning only differences if that makes sense (like the case of OpenSCAP that needs the entire FS for a scan), so only scanners that care about scanning only differences will notice and follow that option (like the clam one).

And for the clam scanner he'll come up with a way to mount once and iterate over and scan only the changed files to avoid copies of potentially large log files, etc., that may be present in long-running containers.

**Comment by legionus** [comment]
@joelsmith @pweil- PTAL again. Added optional file filter. Also root filesystem mounted with `ro`.

Works only with https://github.com/openshift/clam-scanner/pull/4

**Comment by joelsmith** [comment]
@legionus , this looks great.

As it turns out, the `ro` options prevents remounting for DM rootfs devices. I tried several things but could never get one that would give the read-only behavior on a fs that we remount from a DM device. So in my opinion, we should go back to how you had it before I suggested that you add the `ro` option. Sorry about that. ☹️

Also, I'm not sure why I thought that `remount` was necessary, but I think that I somehow confused the needed mount options. Sorry if that wasted your time investigating.

Overall, I think this is ready to go except for the needed removal of the `ro`option.  I tested with one that had `ro` removed again, and it's doing exactly what we want, at least for all the storage drivers I tried (I tried btrfs and device mapper). If you've tested overlay and that's also working, I think we're in good shape.

**Comment by joelsmith** [comment]
Here's some sample output from a container scan:

```json
{
    "ContainerID": "e34748d7f410100b4bfa6622dc3bbf414be8a23ef85532247d974cb5b010de4a",
    "apiVersion": "v1alpha",
    "imageName": "",
    "results": [
        {
            "description": "Eicar-Test-Signature FOUND",
            "name": "clamav",
            "reference": "file:///rootfs/tmp/eicar",
            "scannerVersion": "0.99.2",
            "timestamp": "2017-08-01T17:31:10.236071304Z"
        }
    ]
}
```

Should the `ContainerID` field be defined like this so that the format is consistent fields?

```
ContainerID string `json:"containerID,omitempty"`
```

Also, a question for the wider audience: should `imageName` be `omitempty`, or do we like it present and empty on a container scan?

**Comment by ilackarms** [comment]
we are now using `rancher/trash` to manage dependencies. after you rebase, you'll want to delete `Gopkg.*`, clean out your `vendor` folder.

Add the necessary packages and the needed version/checksum (if any) to `vendor.conf`, then run `trash`

**Comment by legionus** [comment]
@ilackarms Thanks! Done.

**Comment by legionus** [comment]
@joelsmith PTAL

**Comment by joelsmith** [comment]
LGTM. @simon3z, I think we're ready for a review from you. Thanks, everybody, and especially @legionus for helping to get this feature added!

**Comment by simon3z** [comment]
@ilackarms can you check why travis is failing? (Is it an issue in this PR?)

Also AFAIU trash shouldn't have vendored `golang.org/x`. Why is it here? Should @legionus update his PR?

Also, @pweil- @ilackarms can you do another code review?

**Comment by legionus** [comment]
> @ilackarms can you check why travis is failing? (Is it an issue in this PR?)

I know why. I use the [CommandContext](https://golang.org/pkg/os/exec/#CommandContext) to stop openscap execution in case the container is stops and we need to quickly unmount the rootfs. This function is not available in older versions of go, but I did not find another way to interrupt the program.

> Also AFAIU trash shouldn't have vendored golang.org/x. Why is it here? Should @legionus update his PR?

@simon3z  Before that, I got error messages in travis that these modules are not found. So I had to add them.

File: pkg/clamav/scanner.go [hunk 4]
```diff
@@ -8,6 +8,7 @@ import (
 
 	"github.com/fsouza/go-dockerclient"
 	"github.com/openshift/clam-scanner/pkg/clamav"
+	"golang.org/x/net/context"
```
Comment (ilackarms) [inline pkg/clamav/scanner.go]: @simon3z this is why this PR adds `golang.org/x/net`
@legionus is there a reason we're using this package instead of stdlib `context`?

File: pkg/cmd/types_test.go [hunk 5]
```diff
@@ -71,6 +71,10 @@ func TestValidate(t *testing.T) {
 	noSuchPullPolicy.Image = "image"
 	noSuchPullPolicy.PullPolicy = "whatisdocker?"
 
+	conflictOptions := NewDefaultImageInspectorOptions()
+	conflictOptions.Image = "image"
+	conflictOptions.Image = "container"
```
Comment (ilackarms) [inline pkg/cmd/types_test.go]: :+1:

File: pkg/inspector/image-inspector.go [hunk 6]
```diff
@@ -12,14 +12,18 @@ import (
...
 	"strings"
 	"time"
 
 	"archive/tar"
 	"crypto/rand"
 
+	"github.com/containerd/containerd/mount"
 	docker "github.com/fsouza/go-dockerclient"
 	"github.com/openshift/image-inspector/pkg/openscap"
+	"golang.org/x/net/context"
```
Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: same thing about `context` here as in above comment @legionus. just wondering if there's a reason we're preferring golang.org/x's version of context. i'm not sure if there are any differences

File: pkg/inspector/image-inspector.go [hunk 7]
```diff
@@ -124,36 +238,118 @@ func (i *defaultImageInspector) Inspect() error {
...
 
-	imageMetaBefore, inspectErrBefore := client.InspectImage(i.opts.Image)
-	if i.opts.PullPolicy == iiapi.PullNever && inspectErrBefore != nil {
-		return fmt.Errorf("Image %s is not available and pull-policy %s doesn't allow pulling",
-			i.opts.Image, i.opts.PullPolicy)
-	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if len(i.opts.Container) == 0 {
```
Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: nit: this is a very large `if/else` block. i would prefer to extract the contents into functions to make the code a bit more readable (e.g. `extractContainer` and `extractImage` methods; or something along those lines)

**Comment by ilackarms** [comment]
@legionus a general comment about the use of `context` in this PR:

this PR is already a pretty sizeable change to the code. adding `context` changes quite a lot of code as well, including function signatures in many places. is `context` necessary to have as a part of this PR? If not, maybe you can separate those changes into another PR, to reduce the size of this one, and make it easier to evaluate those changes separately?

File: pkg/clamav/scanner.go (hunk 4 above)
Comment (legionus) [reply pkg/clamav/scanner.go]: > @legionus is there a reason we're using this package instead of stdlib context?

@ilackarms OK. I will use the `context` from stdlib.

File: pkg/inspector/image-inspector.go (hunk 6 above)
Comment (legionus) [reply pkg/inspector/image-inspector.go]: @ilackarms There are no differences. I'll use the context from stdlib.

File: pkg/inspector/image-inspector.go (hunk 7 above)
Comment (legionus) [reply pkg/inspector/image-inspector.go]: @ilackarms I can't move it to the function. You missed the fact that I use [defer](https://github.com/openshift/image-inspector/pull/66/files/10875e1d0aafe6a73313b49fdc6e5244da56f67e#diff-cb14c05f76ccda787aced81d2a624e69R328):

```
defer mount.Unmount(i.opts.DstPath, unix.MNT_FORCE|unix.MNT_DETACH)
```

After completing the scan or any error, we must unmount the container rootfs. The context here is used to interrupt all the operations in the case of stopping the container. Therefore, it is created here and is passed to all functions.

**Comment by legionus** [comment]
@ilackarms The use of `context` is the only way to interrupt the execution of operations. For example, without it you cannot interrupt the execution of openscap and unmount the rootfs.

File: pkg/inspector/image-inspector.go (hunk 7 above)
Comment (legionus) [reply pkg/inspector/image-inspector.go]: > ~~I can't move it to the function.~~

@ilackarms I'll make a wrapper which will reduce the size of the `Inspect()` and hide code related to the container filesystem.

**Comment by legionus** [comment]
I found a big issue in this PR.

**Comment by ilackarms** [comment]
😱😱😱

**Comment by ilackarms** [comment]
what's the issue? was it fixed?

it's hard to identify until the CI can run

**Comment by ilackarms** [comment]
also, @legionus i wanted to update this PR with a request: can you please extract all the changes including `context` and cancellation to a separate PR? it's a logically separate set of changes, and useful outside the scope of this PR.

**Comment by legionus** [comment]
@joelsmith @simon3z  @ilackarms  please review

I found a way how to avoid mounting at all. The code has become much simpler. It also allows you to check the mounted volumes without having to copy the content.

**Comment by joelsmith** [comment]
Wow, I had no idea about `/proc/PID/root/`. That's fantastic! I checked, and the kernels we ship with RHEL7 have support for that feature, so it will work for the Ops team. Nice job! 👍

**Comment by simon3z** [comment]
@legionus I think this can be now rebased

**Comment by legionus** [comment]
@simon3z @joelsmith tests green. please review.

**Comment by simon3z** [comment]
> @simon3z @joelsmith tests green. please review.

cc @pweil-

**Comment by joelsmith** [comment]
👍 LGTM

File: pkg/inspector/image-inspector.go [hunk 8]
```diff
@@ -127,36 +133,76 @@ func (i *defaultImageInspector) Inspect() error {
...
-		return err
-	}
+		i.meta.Image = *meta.Image
+		scanResults.ImageID = meta.Image.ID
+		scanResults.ContainerID = meta.Container.ID
 
-	imageMetadata, err := i.createAndExtractImage(client, randomName)
-	if err != nil {
-		return err
+		var filterInclude map[string]struct{}
```
Comment (ilackarms) [inline pkg/inspector/image-inspector.go]: wouldn't this be better as a `map[string]bool`?

Comment (joelsmith) [reply pkg/inspector/image-inspector.go]: @ilackarms I view them as basically equivalent (assuming that values are only ever set to true), but the map[string]struct{} uses less space. I've seen the empty struct idiom used frequently when a string set is wanted. This thread talks about it: https://groups.google.com/forum/#!topic/golang-nuts/lb4xLHq7wug

File: pkg/openscap/openscap.go [hunk 9]
```diff
@@ -191,7 +191,7 @@ func (s *defaultOSCAPScanner) oscapChroot(ctx context.Context, oscapArgs ...stri
 	return out, err
 }
 
-func (s *defaultOSCAPScanner) Scan(ctx context.Context, mountPath string, image *docker.Image) ([]iiapi.Result, interface{}, error) {
+func (s *defaultOSCAPScanner) Scan(ctx context.Context, mountPath string, image *docker.Image, filter iiapi.FilesFilter) ([]iiapi.Result, interface{}, error) {
```
Comment (ilackarms) [inline pkg/openscap/openscap.go]: what's the purpose of having `filter` here? i recall the discussion here https://github.com/openshift/image-inspector/pull/72#discussion_r139773505

if `filter` is not used by all of the scanners, it shouldn't be a part of the interface. thoughts @pweil- ?

Comment (pweil-) [reply pkg/openscap/openscap.go]: > if filter is not used by all of the scanners, it shouldn't be a part of the interface. thoughts @pweil- ?

yes, I do not like it here.  I don't have a solution in mind but it feels like the wrong way to use the interface.

File: pkg/inspector/image-inspector.go (hunk 8 above)
Comment (ilackarms) [reply pkg/inspector/image-inspector.go]: personal preference: I prefer a little better readability over a little better performance

File: pkg/openscap/openscap.go (hunk 9 above)
Comment (legionus) [reply pkg/openscap/openscap.go]: @pweil- @ilackarms The filter is not used here. This is an optional parameter that can narrow the context of the review. If the scanner does not support such a restriction, then it can be ignored.

Comment (joelsmith) [reply pkg/openscap/openscap.go]: IMO, there is nothing wrong with passing the filter to all scanners and allowing every scanner to decide whether to make use of it. Scan types that only operate in a holistic fashion might ignore it while file-wise scanner should use it.

Comment (pweil-) [reply pkg/openscap/openscap.go]: devil's advocate: out of the scanners we have today this works for 50% of them.  We've talked about scanning memory in the past too.  

Maybe what I'm getting at here is that rather than extending the scan interface signature endlessly we make a `ScannerOptions` object.  Then we're in the situation of supporting every possible option for supported scanners in that object which isn't really any different but makes the interface nice.

I think there also might be a pattern for `ScanOptions.Clam` and `ScanOptions.OpenSCAP` which allows more separation and validation on the scanner side.  But we still need a way to pass those back to the scanners from the cli....which then makes me start thinking we need to be able to pass these as a file.

tl;dr I want the interface to be short and the options to be scanner specific if possible

Comment (ilackarms) [reply pkg/openscap/openscap.go]: @pweil- i would go a step further and argue that scanner-specific options should only be passed to 

imo, the purpose of using interfaces in the first place is to decouple the function that calls `scanner.Scan` from the concrete implementation of the scanner. if the caller needs to be aware of `ScanOptions.{Clam,OpenSCAP}`, doesn't that defeat the purpose of using an interface? i think only the section of code that's coupled with the concrete scanner (e.g. the concrete constructor) should be responsible for configuring scanner-specific options. 

in other words, i would suggest adding `ScanOptions` to the struct fields and keep them separate from the interface.

Comment (pweil-) [reply pkg/openscap/openscap.go]: The caller must populate the options at some point.  Whether that's hydrating a type from a file or having command line options.  The reason I like it in the `Scan` interface is that it would work nicely for a persistent scanner where `Scan` can be called multiple times with different options.  I think that's the mental line I draw between what would be `ScanOptions` and `ScannerOptions`.  

This is a good topic for arch but in the spirit of getting this in for the next milestone should we:

1.  add this to the interface
2.  add this to a ScanOptions on the interface
3.  add this to the Scanner instantiation itself

Realistically, we don't have external implementations of the Scanner interface so we're not backing ourselves into a corner either way at this point.  

Preferences or other opinions on this @simon3z @enoodle @mfojtik ?

Comment (pweil-) [reply pkg/openscap/openscap.go]: no one?  Well if that's the case then I think we should leave it as is until we agree on a solution.

**Comment by pweil-** [comment]
@mfojtik review?

**Comment by simon3z** [comment]
@joelsmith @ilackarms @pweil- @mfojtik I am extending the milestone of a week. Can we try to hash out this PR by that? Please check the "reviewers" status, it seems that this PR still has comments and requests of changes to be handled.

File: pkg/openscap/openscap.go (hunk 9 above)
Comment (ilackarms) [reply pkg/openscap/openscap.go]: @pweil- i am fine with leaving as is

File: pkg/cmd/types_test.go (hunk 5 above)
Comment (pweil-) [inline pkg/cmd/types_test.go]: this should be `.Container`?

Comment (legionus) [reply pkg/cmd/types_test.go]: @pweil- good catch! fixed.

**Comment by mfojtik** [comment]
LGTM as well
